master/internal/stream/system.go

```diff
@@ -25,6 +33,10 @@ type JsonB interface{}
 type PublisherSet struct {
 	Trials *stream.Publisher[*TrialMsg]
 	// Experiments *stream.Publisher[*ExperimentMsg]
+
+	closedChan           chan struct{}
```

Comment:
I challenge you to find another way to express whatever it is you need to express here.

My concerns with the `closedChan` as it is currently written are:
- it's confusing why there's multiple shutdown signals that go from the publisher to the streamers.  The streamer is selecting on both the closedChan as well as the permissionChangeChan.  Why isn't one close signal enough?
- it's confusing who is responsible for closing the `closedChan`.  Is it something the publisher closes and others react to?  Or is it something others close and the publisher reacts to?  Right now the only `close(ps.closedChan)` I see is in the `permissionChangeLoop()`.  And yet, the `permissionChangeLoop()` is also watching the `closedChan` for a close signal.  That's a recipe for very complex shutdown ordering bugs.

My recommendation is actually to have a justBootEm function on the PublisherSet, and let the `permissionChangeLoop()` be a standalone function, separate from the PublisherSet (I think... not 100% sure).  But that would make it trivial to unit-test the justBootEm functionality.

My other recommendation is to rely on the context.Context for cancelation-from-above.  Don't have the permissionChangeLoop ask the publisher if the publisher is dead.  Instead feed one context.Context to the whole publishing infrastructure.  There are several subcomponents in that publishing infrastructure, but they shouldn't ask each other if the other is dead; a subcomponent that dies should bubble up to some supervisor, and the common parent can just preempt the whole rest of the publishing infrastructure by canceling the context.Context.  For now, we don't have that supervisor anywhere in the master (for example, if the resource manager dies, the master just sits there useless until it is manually restarted), so you don't need to write that supervisor here.  For now, just rely on the context.Context so that writing such a supervisor later is easy and possible.

Reply:
I'll put some thought into pulling `permissionChangeLoop()` outside of the `PublisherSet`. We could either take the `PublisherSet` created in `core.go` as a parameter as instead of having `permissionChangeLoop()` be a method of PublisherSet.

I had added the `closedChan` in order to respond errors that had previously been ignored in the `permissionChangeLoop()`; however, if we're not expecting there to be a supervisor for managing these failures and requiring a manual restart, then are you suggesting instead to close `permissionChangeChan` and not replace the value, so when a streamer attempts to connect, it instantly gets booted in the publishLoop?

Reply:
Hm, good question.

I think I'd like to lean into the pattern of "crashes are propagated upwards as errors, and supervisors are responsible for canceling the relevant peers of whatever crashed via context.Contexts".  I think that's what context.Contexts are for.

I can see this is looking more like the closedChan that you had before, especially in the "today" world that there is no supervisor.

I think I'd like to see the following supervision hierarchy, basically:

```
         core.go
            |
  publishing supervisor
       /         \
streamer(s)     PublisherSet
                /          \
            publishLoop  permissionChangeLoop
```

Now, any supervising parent hands all its children a context.Context, and monitors all of its children.  If a child dies, it cancels the context.Context it gave out, and passes its error up to its parent.

And now that I draw it out, I don't see that the PublisherSet object really needs to have a monitor loop if a publishing supervisor existed.  The publishing supervisor could easily just monitor the publishLoops and the permissionChangeLoop directly.  And then, why not just have the PublisherSet _be_ the publishing supervisor?  Then you'd want something analgous to the closedChan that you had.  The only differences really are:
- I think it should be a context.Context instead
- I think failures in publishLoop and permissionChangeLoop can't be allowed to cancel the context directly; they should just pass errors to the supervisor and the supervisor would cancel things.
- You would need a supervisor goroutine that detects the errors and cancels the context for the other children.

But the streamer would listen for that context to be closed just like you had it doing with closedChan I think.

Let me know if that's as clear as mud, and we can just have a meeting to discuss further.